File: frontend/src/services/schoolFeesService.ts
import { supabase } from './supabase'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8001/api'

export interface SchoolFee {
  id: string
  grade: string
  annual_fee: number
  term_fee: number
  registration_fee: number
  re_registration_fee: number
  sport_fee: number
  created_at: string
  updated_at: string
}

/**
 * Fetch school fees for a specific grade from backend API
 */
export const getFeeByGrade = async (grade: string): Promise<SchoolFee | null> => {
  try {
    console.log(`Fetching fees for grade: ${grade} from ${API_BASE_URL}`)
    
    const response = await fetch(`${API_BASE_URL}/school-fees/${grade}`)
    
    if (!response.ok) {
      console.error(`Error fetching fees for grade ${grade}: ${response.status} ${response.statusText}`)
      return null
    }

    const data = await response.json()
    console.log(`Fees fetched successfully for grade ${grade}:`, data)
    return data as SchoolFee
  } catch (err) {
    console.error('Unexpected error fetching fees:', err)
    return null
  }
}

/**
 * Fetch all school fees from backend API
 */
export const getAllFees = async (): Promise<SchoolFee[]> => {
  try {
    console.log(`Fetching all fees from ${API_BASE_URL}`)
    
    const response = await fetch(`${API_BASE_URL}/school-fees`)
    
    if (!response.ok) {
      console.error(`Error fetching all fees: ${response.status} ${response.statusText}`)
      return []
    }

    const data = await response.json()
    return (data.fees as SchoolFee[]) || []
  } catch (err) {
    console.error('Unexpected error fetching all fees:', err)
    return []
  }
}

File: backend/services/school_fees_service.py
from typing import Optional, Dict
from core.supabase_client import get_supabase_client


class SchoolFeesService:
    """Service for handling school fees operations"""

    @staticmethod
    async def get_fee_by_grade(grade: str) -> Optional[Dict]:
        """
        Fetch school fees for a specific grade
        
        Args:
            grade: Grade identifier (e.g., 'GR_R', 'GR_1-6', 'GR_7-9', 'GR_10-11', 'GR_12')
            
        Returns:
            Dictionary with fee data or None if not found
        """
        try:
            supabase = get_supabase_client()
            
            # Query without .single() to avoid errors when no rows found
            response = supabase.table('school_fees').select('*').eq('grade', grade).execute()
            
            if response.data and len(response.data) > 0:
                data = response.data[0]
                return {
                    'grade': data.get('grade'),
                    'annual_fee': data.get('annual_fee'),
                    'term_fee': data.get('term_fee'),
                    'registration_fee': data.get('registration_fee'),
                    're_registration_fee': data.get('re_registration_fee'),
                    'sport_fee': data.get('sport_fee', 0)
                }
            return None
            
        except Exception as e:
            print(f"Error fetching fees for grade {grade}: {str(e)}")
            return None

    @staticmethod
    async def get_all_fees() -> list:
        """
        Fetch all school fees
        
        Returns:
            List of fee dictionaries
        """
        try:
            supabase = get_supabase_client()
            
            response = supabase.table('school_fees').select('*').execute()
            
            return response.data if response.data else []
            
        except Exception as e:
            print(f"Error fetching all fees: {str(e)}")
            return []
